File: libs/fleet/sdk-bindings/swift/tests/TestAsyncClient.swift
import Foundation

private struct Expected: @unchecked Sendable {
    let method: String
    let url: String
    let headers: [HttpHeader]
    let body: Data?
    let status: UInt16
    let response: Data
}

private let jsonHeaders = [
    HttpHeader(name: "accept", value: "application/json"),
    HttpHeader(name: "content-type", value: "application/json"),
    HttpHeader(name: "authorization", value: "Bearer offline-token"),
]

private func tokenExpected() -> Expected {
    Expected(
        method: "POST",
        url: "https://keycloak.invalid/token",
        headers: [
            HttpHeader(name: "accept", value: "application/json"),
            HttpHeader(name: "content-type", value: "application/x-www-form-urlencoded"),
            HttpHeader(name: "authorization", value: "Basic Y2xpZW50LWlkOmNsaWVudC1zZWNyZXQ="),
        ],
        body: Data("grant_type=client_credentials".utf8),
        status: 200,
        response: Data(#"{"access_token":"offline-token","expires_in":3600}"#.utf8)
    )
}

private func serviceExpected(body: Data?, response: Data) -> Expected {
    Expected(
        method: "POST",
        url: "https://cyclops.invalid/api/svc/default/offline-sandbox-mcp/mcp",
        headers: [HttpHeader(name: "X-Cua-Fleet-Claim", value: "default"), HttpHeader(name: "authorization", value: "Bearer offline-token")],
        body: body,
        status: 202,
        response: response
    )
}

private func lifecycleQueue() -> [Expected] {
    [
        tokenExpected(),
        Expected(method: "POST", url: "https://cyclops.invalid/api/namespaces", headers: jsonHeaders, body: Data(#"{"name":"default"}"#.utf8), status: 201, response: Data("{}".utf8)),
        Expected(method: "POST", url: "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxwarmpools", headers: jsonHeaders, body: Data(#"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxWarmPool","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"replicas":1,"sandboxTemplateRef":{"name":"default"}},"status":null}"#.utf8), status: 201, response: Data(#"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxWarmPool","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"replicas":1,"sandboxTemplateRef":{"name":"default"}},"status":null}"#.utf8)),
        Expected(method: "POST", url: "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims", headers: jsonHeaders, body: Data(#"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxClaim","metadata":{"namespace":"default","name":"claim-1","labels":null},"spec":{"sandboxTemplateRef":{"name":"default"},"bindDeadline":900},"status":null}"#.utf8), status: 201, response: Data(#"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxClaim","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"sandboxTemplateRef":{"name":"default"}},"status":{"phase":"Bound","sandbox":{"name":"offline-sandbox"}}}"#.utf8)),
        Expected(method: "GET", url: "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default", headers: jsonHeaders, body: nil, status: 200, response: Data(#"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxClaim","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"sandboxTemplateRef":{"name":"default"}},"status":{"phase":"Bound","sandbox":{"name":"offline-sandbox"}}}"#.utf8)),
        Expected(method: "GET", url: "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxtemplates/default", headers: jsonHeaders, body: nil, status: 200, response: Data(#"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxTemplate","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"vmTemplate":{"containerDiskImage":"registry.example/desktop:offline","services":[{"name":"mcp","targetPort":8080}]}}}"#.utf8)),
        serviceExpected(body: Data(#"{"offline":true}"#.utf8), response: Data("offline service accepted".utf8)),
        Expected(method: "DELETE", url: "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default", headers: jsonHeaders, body: nil, status: 204, response: Data()),
        Expected(method: "DELETE", url: "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxwarmpools/default", headers: jsonHeaders, body: nil, status: 204, response: Data()),
        Expected(method: "DELETE", url: "https://cyclops.invalid/api/namespaces/default", headers: jsonHeaders, body: nil, status: 204, response: Data()),
    ]
}

private func normalizedGeneratedClaimBody(_ body: Data?) -> Data? {
    guard let body,
          var object = try? JSONSerialization.jsonObject(with: body) as? [String: Any],
          var metadata = object["metadata"] as? [String: Any],
          let name = metadata["name"] as? String,
          name.hasPrefix("claim-") else { return nil }
    let suffix = name.dropFirst("claim-".count)
    guard !suffix.isEmpty,
          suffix.first != "-",
          suffix.last != "-",
          name.utf8.count <= 63,
          suffix.allSatisfy({ $0.isLowercase || $0.isNumber || $0 == "-" }) else { return nil }
    metadata["name"] = "claim-generated"
    object["metadata"] = metadata
    return try? JSONSerialization.data(withJSONObject: object, options: [.sortedKeys])
}

actor ScriptedHttpClient: HttpClient {
    private var expected: [Expected]
    private var requests: [HttpRequest] = []

    fileprivate init(expected: [Expected]) {
        self.expected = expected
    }

    func execute(request: HttpRequest) async throws -> HttpResponse {
        guard !expected.isEmpty else { preconditionFailure("unexpected request") }
        let item = expected.removeFirst()
        precondition(request.method == item.method && request.url == item.url)
        precondition(request.headers == item.headers)
        if request.method == "POST" && request.url.hasSuffix("/osgymsandboxclaims") {
            precondition(normalizedGeneratedClaimBody(request.body) == normalizedGeneratedClaimBody(item.body))
        } else {
            precondition(request.body == item.body)
        }
        requests.append(request)
        return HttpResponse(status: item.status, headers: [], body: item.response)
    }

    func assertExhausted() { precondition(expected.isEmpty) }
    func recordedRequestCount() -> Int { requests.count }
}

final class FailingHttpClient: HttpClient, @unchecked Sendable {
    func execute(request: HttpRequest) async throws -> HttpResponse {
        throw HttpError.Transport(reason: "scripted callback failure")
    }
}

private func configuration() -> CyclopsConfiguration {
    CyclopsConfiguration(baseUrl: "https://cyclops.invalid", tokenUrl: "https://keycloak.invalid/token", credentials: CyclopsCredentials(clientId: "client-id", clientSecret: "client-secret"), poolPollIntervalMs: 1, poolPollLimit: 1, claimPollIntervalMs: 1, claimPollLimit: 2)
}

private let sandbox = Sandbox(namespace: "default", claim: "default", name: "offline-sandbox", services: ["mcp"])
private func serviceRequest(_ body: Data?) -> HttpRequest {
    HttpRequest(method: "POST", url: "https://ignored.invalid/mcp", headers: [], body: body, timeoutSecs: nil)
}

@main struct AppControlled {
    static func main() async throws {
        let spec = OsGymSandboxWarmPoolSpec(replicas: 1, sandboxTemplateRef: SandboxTemplateRef(name: "default"), autoscaling: nil)
        let transport = ScriptedHttpClient(expected: lifecycleQueue())
        let client = try CyclopsClient.connect(configuration: configuration(), httpClient: transport)
        let pool = try await client.createPool(request: CreatePoolRequest(namespace: "default", spec: spec))
        let claim = try await client.createClaim(request: CreateClaimRequest(pool: pool, spec: ClaimSpec(sandboxTemplateRef: SandboxTemplateRef(name: pool.metadata.name), warmpool: nil, bindDeadline: nil, lifecycle: nil)))
        let createdSandbox = try await client.waitClaim(claim: claim)
        let service = try await client.serviceRequest(sandbox: createdSandbox, service: "mcp", path: "/mcp", request: serviceRequest(Data(#"{"offline":true}"#.utf8)))
        try await client.deleteClaim(claim: claim)
        try await client.deletePool(pool: pool)
        await transport.assertExhausted()
        precondition(service.status == 202)

        let failingClient = try CyclopsClient.connect(configuration: configuration(), httpClient: FailingHttpClient())
        do {
            _ = try await failingClient.serviceRequest(sandbox: sandbox, service: "mcp", path: "/mcp", request: serviceRequest(nil))
            preconditionFailure("expected generated SdkError.Transport")
        } catch SdkError.Transport(let reason) {
            precondition(reason == "scripted callback failure")
        }

        let bodyTransport = ScriptedHttpClient(expected: [
            tokenExpected(),
            serviceExpected(body: nil, response: Data()),
            serviceExpected(body: Data(), response: Data([0, 0x7f, 0xff])),
        ])
        let bodyClient = try CyclopsClient.connect(configuration: configuration(), httpClient: bodyTransport)
        let emptyBodyResponse = try await bodyClient.serviceRequest(sandbox: sandbox, service: "mcp", path: "/mcp", request: serviceRequest(nil))
        precondition(emptyBodyResponse.body.isEmpty)
        let binaryBodyResponse = try await bodyClient.serviceRequest(sandbox: sandbox, service: "mcp", path: "/mcp", request: serviceRequest(Data()))
        precondition(binaryBodyResponse.body == Data([0, 0x7f, 0xff]))
        await bodyTransport.assertExhausted()

        let concurrentTransport = ScriptedHttpClient(expected: [
            tokenExpected(),
            serviceExpected(body: nil, response: Data("warm".utf8)),
            serviceExpected(body: Data("parallel".utf8), response: Data("first".utf8)),
            serviceExpected(body: Data("parallel".utf8), response: Data("second".utf8)),
        ])
        let concurrentClient = try CyclopsClient.connect(configuration: configuration(), httpClient: concurrentTransport)
        _ = try await concurrentClient.serviceRequest(sandbox: sandbox, service: "mcp", path: "/mcp", request: serviceRequest(nil))
        async let first = concurrentClient.serviceRequest(sandbox: sandbox, service: "mcp", path: "/mcp", request: serviceRequest(Data("parallel".utf8)))
        async let second = concurrentClient.serviceRequest(sandbox: sandbox, service: "mcp", path: "/mcp", request: serviceRequest(Data("parallel".utf8)))
        let concurrentResponses = try await (first, second)
        precondition(Set([concurrentResponses.0.body, concurrentResponses.1.body]) == Set([Data("first".utf8), Data("second".utf8)]))
        await concurrentTransport.assertExhausted()
        let recordedRequestCount = await concurrentTransport.recordedRequestCount()
        precondition(recordedRequestCount == 4)
        print("pool=\(pool.metadata.name) claim=\(claim.metadata.name) sandbox=\(createdSandbox.name) service_status=\(service.status)")
    }
}

File: libs/fleet/sdk-bindings/ruby/tests/test_async_client.rb
require 'json'
require 'cyclops_sdk'

Expected = Struct.new(:method, :url, :headers, :body, :status, :response)
BASE = 'https://cyclops.invalid'
TOKEN = 'https://keycloak.invalid/token'
JSON_HEADERS = [['accept', 'application/json'], ['content-type', 'application/json'], ['authorization', 'Bearer offline-token']]

GENERATED_CLAIM_NAME = /\Aclaim-[a-z0-9](?:[-a-z0-9]*[a-z0-9])?\z/

def normalized_generated_claim_body(body)
  return nil if body.nil?
  value = JSON.parse(body)
  name = value.fetch('metadata').fetch('name')
  raise "invalid generated claim name: #{name.inspect}" unless name.bytesize <= 63 && GENERATED_CLAIM_NAME.match?(name)
  value['metadata']['name'] = 'claim-generated'
  JSON.generate(value).b
end

def pool_json
  { apiVersion: 'osgym.cua.ai/v1alpha1', kind: 'OSGymSandboxWarmPool', metadata: { namespace: 'default', name: 'default', labels: nil }, spec: { replicas: 1, sandboxTemplateRef: { name: 'default' } }, status: nil }
end

def template_json
  { apiVersion: 'osgym.cua.ai/v1alpha1', kind: 'OSGymSandboxTemplate', metadata: { namespace: 'default', name: 'default', labels: nil }, spec: { vmTemplate: { containerDiskImage: 'registry.example/desktop:offline', services: [{ name: 'mcp', targetPort: 8080 }] } } }
end

def claim_json(bound = false)
  value = { apiVersion: 'osgym.cua.ai/v1alpha1', kind: 'OSGymSandboxClaim', metadata: { namespace: 'default', name: 'default', labels: nil }, spec: { sandboxTemplateRef: { name: 'default' } }, status: nil }
  value[:status] = { phase: 'Bound', sandbox: { name: 'offline-sandbox' } } if bound
  value
end

def token_expected
  Expected.new('POST', TOKEN, [['accept', 'application/json'], ['content-type', 'application/x-www-form-urlencoded'], ['authorization', 'Basic Y2xpZW50LWlkOmNsaWVudC1zZWNyZXQ=']], 'grant_type=client_credentials'.b, 200, { access_token: 'offline-token', expires_in: 3600 })
end

def service_expected(body, response)
  Expected.new('POST', "#{BASE}/api/svc/default/offline-sandbox-mcp/mcp", [['X-Cua-Fleet-Claim', 'default'], ['authorization', 'Bearer offline-token']], body, 202, response)
end

def lifecycle_queue
  pool_url = "#{BASE}/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxwarmpools/default"
  template_url = "#{BASE}/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxtemplates/default"
  claim_url = "#{BASE}/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default"
  pool_body = JSON.generate(pool_json).b
  claim_body = '{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxClaim","metadata":{"namespace":"default","name":"claim-1","labels":null},"spec":{"sandboxTemplateRef":{"name":"default"},"bindDeadline":900},"status":null}'.b
  [
    token_expected,
    Expected.new('POST', "#{BASE}/api/namespaces", JSON_HEADERS, '{"name":"default"}'.b, 201, {}),
    Expected.new('POST', pool_url.sub(%r{/default\z}, ''), JSON_HEADERS, pool_body, 201, pool_json),
    Expected.new('POST', claim_url.sub(%r{/default\z}, ''), JSON_HEADERS, claim_body, 201, claim_json),
    Expected.new('GET', claim_url, JSON_HEADERS, nil, 200, claim_json(true)),
    Expected.new('GET', template_url, JSON_HEADERS, nil, 200, template_json),
    service_expected('{"offline":true}'.b, 'offline service accepted'),
    Expected.new('DELETE', claim_url, JSON_HEADERS, nil, 204, ''),
    Expected.new('DELETE', pool_url, JSON_HEADERS, nil, 204, ''),
    Expected.new('DELETE', "#{BASE}/api/namespaces/default", JSON_HEADERS, nil, 204, ''),
  ]
end

class ScriptedHttpClient < FleetSdk::HttpClient
  def initialize(expected)
    @expected = expected
    @mutex = Mutex.new
    @requests = []
  end

  def execute(request)
    @mutex.synchronize do
      item = @expected.shift or raise 'unexpected request'
      actual = [request.method, request.url, request.headers.map { |header| [header.name, header.value] }, request.body]
      expected = [item.method, item.url, item.headers, item.body]
      if request.method == 'POST' && request.url.end_with?('/osgymsandboxclaims')
        raise "request mismatch: #{actual.inspect}" unless actual.first(3) == expected.first(3) && normalized_generated_claim_body(request.body) == normalized_generated_claim_body(item.body)
      else
        raise "request mismatch: #{actual.inspect}" unless actual == [item.method, item.url, item.headers, item.body]
      end
      @requests << request
      body = item.response.is_a?(String) ? item.response.b : JSON.generate(item.response).b
      FleetSdk::HttpResponse.new(status: item.status, headers: [], body: body)
    end
  end

  def assert_exhausted!
    @mutex.synchronize { raise @expected.inspect unless @expected.empty? }
  end

  def request_count
    @mutex.synchronize { @requests.length }
  end
end

class FailingHttpClient < FleetSdk::HttpClient
  def execute(request)
    raise FleetSdk::HttpError::Transport.new(reason: 'scripted callback failure')
  end
end

def configuration
  FleetSdk::CyclopsConfiguration.new(base_url: BASE, token_url: TOKEN, credentials: FleetSdk::CyclopsCredentials.new('client-id', 'client-secret'), pool_poll_interval_ms: 1, pool_poll_limit: 1, claim_poll_interval_ms: 1, claim_poll_limit: 2)
end

def service_request(body)
  builder = FleetSdk::HttpRequestBuilder.new.method('POST').url('https://ignored.invalid/mcp').headers([])
  builder = builder.body(body) unless body.nil?
  builder.build
end

sandbox = FleetSdk::Sandbox.new(namespace: 'default', claim: 'default', name: 'offline-sandbox', services: ['mcp'])
spec = FleetSdk::OSGymSandboxWarmPoolSpec.new(replicas: 1, sandbox_template_ref: FleetSdk::SandboxTemplateRef.new(name: 'default'), autoscaling: nil)
transport = ScriptedHttpClient.new(lifecycle_queue)
client = FleetSdk::CyclopsClient.connect(configuration, transport)
pool = client.create_pool(FleetSdk::CreatePoolRequest.new(namespace: 'default', spec: spec))
claim = client.create_claim(FleetSdk::CreateClaimRequest.new(pool: pool, spec: FleetSdk::ClaimSpec.new(sandbox_template_ref: FleetSdk::SandboxTemplateRef.new(name: pool.metadata.name), warmpool: nil, bind_deadline: nil, lifecycle: nil)))
created_sandbox = client.wait_claim(claim)
service = client.service_request(created_sandbox, 'mcp', '/mcp', service_request('{"offline":true}'.b))
client.delete_claim(claim)
client.delete_pool(pool)
transport.assert_exhausted!
raise 'unexpected service response' unless service.status == 202

failing_client = FleetSdk::CyclopsClient.connect(configuration, FailingHttpClient.new)
begin
  failing_client.service_request(sandbox, 'mcp', '/mcp', service_request(nil))
  raise 'expected generated SdkError::Transport'
rescue FleetSdk::SdkError::Transport => error
  raise error unless error.reason == 'scripted callback failure'
end

body_transport = ScriptedHttpClient.new([
  token_expected,
  service_expected(nil, "".b),
  service_expected("".b, "\x00\x7f\xff".b),
])
body_client = FleetSdk::CyclopsClient.connect(configuration, body_transport)
raise 'expected empty service body' unless body_client.service_request(sandbox, 'mcp', '/mcp', service_request(nil)).body.empty?
raise 'expected binary service body' unless body_client.service_request(sandbox, 'mcp', '/mcp', service_request("".b)).body == "\x00\x7f\xff".b
body_transport.assert_exhausted!

concurrent_transport = ScriptedHttpClient.new([
  token_expected,
  service_expected(nil, 'warm'),
  service_expected('parallel'.b, 'first'),
  service_expected('parallel'.b, 'second'),
])
concurrent_client = FleetSdk::CyclopsClient.connect(configuration, concurrent_transport)
concurrent_client.service_request(sandbox, 'mcp', '/mcp', service_request(nil))
threads = 2.times.map do
  Thread.new do
    concurrent_client.service_request(sandbox, 'mcp', '/mcp', service_request('parallel'.b))
  end
end
responses = threads.map(&:value)
raise 'concurrent SDK calls did not both complete' unless responses.map { |response| response.body }.sort == %w[first second]
concurrent_transport.assert_exhausted!
raise 'concurrent queue recording mismatch' unless concurrent_transport.request_count == 4
puts "pool=#{pool.metadata.name} claim=#{claim.metadata.name} sandbox=#{created_sandbox.name} service_status=#{service.status}"
